src/error/error.rs

```diff
@@ -5,43 +5,80 @@ use std::fmt;
 use http::status::StatusCode;
 
 /// Errors that can happen inside Tower Web.
+/// The object of this type is serializable into "Problem Detail" as defined in RFC7807.
+#[derive(Serialize)]
 pub struct Error {
-    kind: ErrorKind,
-}
-
-/// A list specifying the general categories of Tower Web errors.
-pub struct ErrorKind {
-    kind: KindPriv,
-}
-
-#[derive(Debug, Eq, PartialEq, Clone, Copy)]
-enum KindPriv {
-    BadRequest,
-    Unauthorized,
-    Fordidden,
-    NotFound,
-    UnprocessableEntity,
-    Internal,
+    #[serde(rename = "type")]
+    kind: String,
+    title: String,
+    // NOTE: type of this property might be changed in the future.
+    // Nevertheless, String type should be accepted.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    detail: Option<String>,
+    #[serde(skip)]
+    status: StatusCode,
+
+    // TODO: this property isn't used and should be removed
+    #[serde(skip)]
+    error_kind: ErrorKind,
 }
 
 // ===== impl Error =====
 
 impl Error {
+    // TODO: this function should return '&self.kind' property
+    // and its return value should be changed to '&str'
     /// Returns the corresponding `ErrorKind` for this error.
+    #[deprecated(note="return value of the function will be changed to &str")]
     pub fn kind(&self) -> &ErrorKind {
-        &self.kind
+        &self.error_kind
+    }
+
+    /// Creates a blank error object specifying only a status code without any details.
+    pub fn blank(status: StatusCode) -> Self {
+        let title = status.canonical_reason().unwrap_or("Unknown status code");
+        Self {
+            kind: String::from("about:blank"),
+            title: title.to_owned(),
+            detail: None,
+            status,
+
+            // TODO: this property isn't used and should be removed
+            error_kind: ErrorKind::new(&status),
+        }
+    }
+
+    /// Creates an error object.
+    pub fn new(kind: &str, title: &str, status: StatusCode) -> Self {
+        Self {
+            kind: kind.to_owned(),
+            title: title.to_owned(),
+            detail: None,
+            status,
+
+            // TODO: this property isn't used and should be removed
+            error_kind: ErrorKind::new(&status),
+        }
+    }
+
+    // NOTE: type of this property might be changed in the future.
+    // Nevertheless, String type should be accepted.
+    /// Provides detailed information about the error.
+    pub fn detail(self, detail: &str) -> Self {
```

Comment:
Can this be changed to: `fn set_detail(&mut self, value: &str) { .. }`. I prefer to avoid "builder" style APIs on non-builder types (if that makes sense).

Or the fn can be deleted for the initial PR.

Reply:
This is function is the most important part of the PR :-) I think "builder" style APIs provide less verbose way of creating objects of `Error` type. Maybe it better to rename the type itself (split into two).

Reply:
If we change it to:

```rust
pub fn set_detail(&mut self, value: &str) -> &mut Self {
    self.detail = Some(detail.to_owned());
    self
}
```

It will make impossible to build an `Error` like that:

```rust
object.ok_or_else(|| Error::new("foo", "bar", StatusCode::INTERNAL_SERVER_ERROR).detail("baz"))?,
```

Because of:

```rust
object.ok_or_else(|| *Error::new("foo", "bar", StatusCode::INTERNAL_SERVER_ERROR).set_detail("baz"))?,
```

```rust
error[E0507]: cannot move out of borrowed content
  --> src/app/util.rs:73:32
   |
73 object.ok_or_else(|| *Error::new("foo", "bar", StatusCode::INTERNAL_SERVER_ERROR).set_detail("baz"))?,
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
```

The way it will work seems to me quite verbose

```rust
object.ok_or_else(|| { 
    let mut err = Error::new("foo", "bar", StatusCode::INTERNAL_SERVER_ERROR); 
    err.set_detail("baz");
    err 
})?,
```

Am I missing something?

Reply:
Ok, I'll add the builder type

Reply:
@carllerche I've added ErrorBuilder type. I intentionally avoided returning `Result` from `build()` function by using `builder(status)` and `kind(kind, title)` in order to do not deal with errors of creating errors :-)

Reply:
Is it a good practice to use `Error::builder` instead of `ErrorBuilder::new`? Are there any rules how to chose between these two options?

Reply:
@carllerche could you please take a look? Is there anything else that need to be done?

Reply:
In practice, I've found useful to create a builder with a particular error kind (and title) and then add a status and some detail to constructed error using it. For that I've implemented zero-argument constructor for `error::Builder`. 

In case of a building error, an error with kind=error_builder_error, status=INTERNAL_SERVER_ERROR and detail describing missing arguments will be returned.